src/functor.jl

```diff
@@ -36,16 +36,14 @@ Possible values include:
 """
 trainmode!(m, mode = true) = mode isa Bool ? testmode!(m, !mode) : testmode!(m, mode)
 
+params!(p::Params, x::DenseArray{<:Number}, seen = IdSet()) = Functors.isleaf(x) && push!(p, x)
```

Comment:
Can a DenseArray ever not be a leaf?

Reply:
Technically yes, `Base.Experimental.Const` is a pure wrapper type and subtypes `DenseArray`. I've seen it used in JuliaGPU libraries, but am unsure if those would come anywhere close to `params`.

Reply:
That's an interesting type. But I guess it will always be leaf-like, Functors should treat it as it would an SArray right? 

More broadly if this method has a test for `isleaf`, then it has to do something with the other branch. And then it's the other method. I guess it could assert isleaf just to make sure you get an error if someone does something really weird.

Reply:
I think it has to be a non-leaf for the same reason `Transpose` does: shared inner arrays.

RE the other branch, I thought the latest change addressed that but it appears I misremembered. Silently dropping an array instead of recursing is definitely not good.

Reply:
I guess. Although transposing one of two shared arrays is common, but marking as Const only one of the two seems perverse.

Reply:
At some point Functors was going to check pointers, or unwrap, to try to give an error on sharing it can't handle (like reshape). Surprisingly neither of these work here:
```
julia> parent(Base.Experimental.Const([1,2]))
2-element Base.Experimental.Const{Int64, 1}:
 1
 2

julia> pointer(Base.Experimental.Const([1,2]))
ERROR: conversion to pointer not defined for Base.Experimental.Const{Int64, 1}
```

Reply:
`pointer` has pretty poor coverage and behaves a little weird too (this is what I was thinking about with `SubArray`). `pointer_from_objref` is not much better. I think that's part of why we backed off from trying to be too clever. If people want to use reshaped arrays without explicitly tying them in the future Optimisers.jl interface, caveat emptor.

Reply:
Let me update my suggestion. I think this ought to be safe, and will at least throw an error should someone ever `@functor Base.Experimental.Const ` (or its CUDA analogue):
```suggestion
function params!(p::Params, x::DenseArray{<:Number}, seen)
  # Fast path for the most common case, Array & CuArray. Solves issue 2040.
  Functors.isleaf(x) || error("For efficiency, params believes every DenseArray of numbers is leaflike")
  push!(p, x)
end
```

Reply:
This code returns `size.(Flux.params((x = view([1,2,3]pi, 1:2), y = transpose([4 5]pi)))) == [(1, 2)]`.

Reply:
This code and what else?

Reply:
This code with the above suggestion.

Reply:
Ok. On this example, the suggestion changes nothing compared to the PR. It just moves the `isleaf` test to be an error not an ignore. 

I think such a fast method should exists alongside the method which was here before the PR, which handles all cases (but has more branches). That should be correct. Whether it still solves 2040 I don't know.